soufi/finder.py

```diff
@@ -218,6 +222,46 @@ def find(
             )
         return self._find()
 
+    # Use this wrapper for doing HTTP HEAD requests, as it will swallow
+    # Timeout exceptions, but cache other lookups.
+    def test_url(self, url, **kwargs):
+        try:
+            return self._head_url(url, **kwargs)
+        except requests.exceptions.Timeout:
+            return False
+
+    def _head_url(self, url, **kwargs):
+        def inner(url):
+            response = requests.head(url, timeout=self.TIMEOUT, **kwargs)
+            if response.status_code == requests.codes.not_allowed:
+                # HEAD not available; we can try to download it instead and
+                # abort before starting the stream.
+                response = requests.get(
+                    url, stream=True, timeout=self.TIMEOUT, **kwargs
+                )
+                response.close()
+            return response.status_code == requests.codes.ok
+
+        return self._cache.get_or_create(
+            f"head-{url}", inner, creator_args=([url], {})
+        )
+
+    def get_url(self, url, params=None, **kwargs):
+        def inner(url):
+            response = requests.get(
+                url, params=params, timeout=self.TIMEOUT, **kwargs
+            )
+            if response.status_code != requests.codes.ok:
+                raise exceptions.DownloadError(response.reason)
+            return response
+
+        key = url
+        if params:
+            key += f"?{params}"
```

Comment:
Does this and `_head_url` intentionally ignore kwargs?

Reply:
They don't, and they won't, because `kwargs` will be in-scope for the closure.  But it is confusing as-written; I don't know why I didn't go all-in on using the closure and have it not accept any inner args at all.

Will fix.

Reply:
Yeah I was about to say why use a closure at all, but this works too.